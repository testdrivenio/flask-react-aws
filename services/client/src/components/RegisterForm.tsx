import React from "react";
import { Formik, Form, Field } from "formik";
import { z } from "zod";
import {
  Box,
  Button,
  FormControl,
  FormLabel,
  Input,
  FormErrorMessage,
  VStack,
  Heading,
} from "@chakra-ui/react";
import { Navigate } from "react-router-dom";

interface RegisterFormProps {
  onSubmit: (values: {
    username: string;
    email: string;
    password: string;
  }) => Promise<void>;
  isAuthenticated: () => boolean;
}

const validationSchema = z.object({
  username: z
    .string()
    .min(6, "Username must be at least 6 characters long")
    .regex(
      /^[a-zA-Z0-9_]+$/,
      "Username can only contain letters, numbers, and underscores",
    ),
  email: z
    .string()
    .email("Enter a valid email")
    .min(6, "Email must be at least 6 characters long"),
  password: z.string().min(11, "Password must be at least 11 characters long"),
});

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type FormValues = z.infer<typeof validationSchema>;

const RegisterForm: React.FC<RegisterFormProps> = ({
  onSubmit,
  isAuthenticated,
}) => {
  if (isAuthenticated()) {
    return <Navigate to="/" replace />;
  }

  return (
    <Box maxWidth="400px" marginTop={10} margin="100px auto 0">
      <Heading as="h1" size="xl" textAlign="center" mb={6}>
        Register
      </Heading>
      <Formik
        initialValues={{
          username: "",
          email: "",
          password: "",
        }}
        onSubmit={async (values, { setSubmitting, resetForm }) => {
          try {
            await onSubmit(values);
            resetForm();
          } catch (error) {
            console.error("Registration failed:", error);
          } finally {
            setSubmitting(false);
          }
        }}
        validate={(values) => {
          try {
            validationSchema.parse(values);
            return {};
          } catch (error) {
            return (error as z.ZodError).formErrors.fieldErrors;
          }
        }}
      >
        {({ errors, touched, isSubmitting }) => (
          <Form>
            <VStack spacing={4}>
              <Field name="username">
                {({ field }) => (
                  <FormControl
                    isInvalid={!!(errors.username && touched.username)}
                  >
                    <FormLabel htmlFor="username">Username</FormLabel>
                    <Input
                      {...field}
                      id="username"
                      placeholder="Enter a username"
                    />
                    <FormErrorMessage data-testid="errors-username">
                      {errors.username}
                    </FormErrorMessage>
                  </FormControl>
                )}
              </Field>
              <Field name="email">
                {({ field }) => (
                  <FormControl isInvalid={!!(errors.email && touched.email)}>
                    <FormLabel htmlFor="email">Email</FormLabel>
                    <Input
                      {...field}
                      id="email"
                      type="email"
                      placeholder="Enter an email address"
                    />
                    <FormErrorMessage data-testid="errors-email">
                      {errors.email}
                    </FormErrorMessage>
                  </FormControl>
                )}
              </Field>
              <Field name="password">
                {({ field }) => (
                  <FormControl
                    isInvalid={!!(errors.password && touched.password)}
                  >
                    <FormLabel htmlFor="password">Password</FormLabel>
                    <Input
                      {...field}
                      id="password"
                      type="password"
                      placeholder="Enter a password"
                    />
                    <FormErrorMessage data-testid="errors-password">
                      {errors.password}
                    </FormErrorMessage>
                  </FormControl>
                )}
              </Field>
              <Button
                type="submit"
                colorScheme="blue"
                width="full"
                isLoading={isSubmitting}
                bg="green.400"
                _hover={{ bg: "green.500" }}
              >
                Register
              </Button>
            </VStack>
          </Form>
        )}
      </Formik>
    </Box>
  );
};

export default RegisterForm;
