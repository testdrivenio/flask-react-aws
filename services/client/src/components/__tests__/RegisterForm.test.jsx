import React from 'react';
import { cleanup, fireEvent, wait } from '@testing-library/react';

import RegisterForm from '../RegisterForm';

afterEach(cleanup);


describe('renders', () => {
  const props = {
    handleRegisterFormSubmit: () => { return true },
    isAuthenticated: () => { return false },
  }

  it('properly', () => {
    const { getByText } = renderWithRouter(<RegisterForm {...props} />);
    expect(getByText('Register')).toHaveClass('title');
  });

  it('default props', () => {
    const { getByLabelText, getByText } = renderWithRouter(<RegisterForm {...props} />);

    const usernameInput = getByLabelText('Username');
    expect(usernameInput).toHaveAttribute('type', 'text');
    expect(usernameInput).not.toHaveValue();

    const emailInput = getByLabelText('Email');
    expect(emailInput).toHaveAttribute('type', 'email');
    expect(emailInput).not.toHaveValue();

    const passwordInput = getByLabelText('Password');
    expect(passwordInput).toHaveAttribute('type', 'password');
    expect(passwordInput).not.toHaveValue();

    const buttonInput = getByText('Submit');
    expect(buttonInput).toHaveValue('Submit');
  });

  it("a snapshot properly", () => {
    const { asFragment } = renderWithRouter(<RegisterForm {...props} />);
    expect(asFragment()).toMatchSnapshot();
  });
});

describe('handles form validation correctly', () => {
  const mockProps = {
    handleRegisterFormSubmit: jest.fn(),
    isAuthenticated: jest.fn(),
  };

  it('when fields are empty', async() => {
    const {
      getByLabelText,
      container,
      findByTestId
    } = renderWithRouter(<RegisterForm {...mockProps} />);

    const form = container.querySelector('form');
    const usernameInput = getByLabelText('Username');
    const emailInput = getByLabelText('Email');
    const passwordInput = getByLabelText('Password');

    expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);

    fireEvent.blur(usernameInput);
    fireEvent.blur(emailInput);
    fireEvent.blur(passwordInput);

    expect((await findByTestId('errors-username')).innerHTML).toBe('Username is required.');
    expect((await findByTestId('errors-email')).innerHTML).toBe('Email is required.');
    expect((await findByTestId('errors-password')).innerHTML).toBe('Password is required.');

    fireEvent.submit(form);

    await wait(() => {
      expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);
    });
  });

  it('when email field is not valid', async() => {
    const {
      getByLabelText,
      container,
      findByTestId
    } = renderWithRouter(<RegisterForm {...mockProps} />);

    const form = container.querySelector('form');
    const emailInput = getByLabelText('Email');

    expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);

    fireEvent.change(emailInput, { target: { value: 'invalid' } });
    fireEvent.blur(emailInput);

    expect((await findByTestId('errors-email')).innerHTML).toBe('Enter a valid email.');

    fireEvent.submit(form);

    await wait(() => {
      expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);
    });
  });

  it('when fields are not the proper length', async() => {
    const {
      getByLabelText,
      container,
      findByTestId
    } = renderWithRouter(<RegisterForm {...mockProps} />);

    const form = container.querySelector('form');
    const usernameInput = getByLabelText('Username');
    const emailInput = getByLabelText('Email');
    const passwordInput = getByLabelText('Password');

    expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);

    fireEvent.change(usernameInput, { target: { value: 'null' } });
    fireEvent.change(emailInput, { target: { value: 't@t.c' } });
    fireEvent.change(passwordInput, { target: { value: 'invalid' } });
    fireEvent.blur(usernameInput);
    fireEvent.blur(emailInput);
    fireEvent.blur(passwordInput);

    expect((await findByTestId('errors-username')).innerHTML).toBe('Username must be greater than 5 characters.');
    expect((await findByTestId('errors-email')).innerHTML).toBe('Email must be greater than 5 characters.');
    expect((await findByTestId('errors-password')).innerHTML).toBe('Password must be greater than 10 characters.');

    fireEvent.submit(form);

    await wait(() => {
      expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);
    });
  });

  it('when fields are valid', async() => {
    const {
      getByLabelText,
      container,
      findByTestId
    } = renderWithRouter(<RegisterForm {...mockProps} />);

    const form = container.querySelector('form');
    const usernameInput = getByLabelText('Username');
    const emailInput = getByLabelText('Email');
    const passwordInput = getByLabelText('Password');

    expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(0);

    fireEvent.change(usernameInput, { target: { value: 'proper' } });
    fireEvent.change(emailInput, { target: { value: 't@t.com' } });
    fireEvent.change(passwordInput, { target: { value: 'properlength' } });
    fireEvent.blur(usernameInput);
    fireEvent.blur(emailInput);
    fireEvent.blur(passwordInput);

    fireEvent.submit(form);

    await wait(() => {
      expect(mockProps.handleRegisterFormSubmit).toHaveBeenCalledTimes(1);
    });
  });
});
