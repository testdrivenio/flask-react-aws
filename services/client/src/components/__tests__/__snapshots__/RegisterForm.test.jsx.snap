// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a snapshot properly 1`] = `
<DocumentFragment>
  <div>
    <h1
      class="title is-1"
    >
      Register
    </h1>
    <hr />
    <br />
    <form>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-username"
        >
          Username
        </label>
        <input
          class="input"
          id="input-username"
          name="username"
          placeholder="Enter a username"
          type="text"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-email"
        >
          Email
        </label>
        <input
          class="input"
          id="input-email"
          name="email"
          placeholder="Enter an email address"
          type="email"
          value=""
        />
      </div>
      <div
        class="field"
      >
        <label
          class="label"
          for="input-password"
        >
          Password
        </label>
        <input
          class="input"
          id="input-password"
          name="password"
          placeholder="Enter a password"
          type="password"
          value=""
        />
      </div>
      <input
        class="button is-primary"
        type="submit"
        value="Submit"
      />
    </form>
  </div>
</DocumentFragment>
`;
